### DISCUSSION TITLE
Entities with multiple keys

### DISCUSSION THREAD

**Discussion by arcanis** [urql-graphql/urql#2519] [Q&A] [answered]
My `User` entities have a typical `id` field, but also a `username` one. Both can be used to query a user (through the `Query#user` endpoint for `id`-based selections, and `Query#userByUsername` for `username`-based ones). I typically only use the first, except in the important case of page loading: the page url contains the username, not the ID.

When switching from a page to another in my app, if `Query#userByUsername` isn't in cache already, the page briefly blinks - even if the user is already in memory. I thought this would be solved by adding a local resolver telling `urql` that `username` is a key, but it didn't work:

```js
cacheExchange({
  resolvers: {
    Query: {
      user: (_, args) => ({ __typename: 'User', id: args.id }),
      userByUsername: (_, args) => ({ __typename: 'User', username: args.username }),
    },
  },
});
```

I'd expect this syntax to instruct urql to cache the same entity under two cache keys (assuming the username is part of all my queries, of course), but instead it seems to break some expectation and the query just constantly invalidate itself.

Is it supposed to be supported? Or would it be a feature request?

**Comment by kitten** [comment, marked as answer]
It's a feature request; unfortunately, we currently do not support secondary keys, which would require a second index. This is pretty tricky as well as it adds a lot of complexity and makes the config potentially harder to understand.

I wonder though whether we should finally allow the cache to give a list of all entities of a type it knows, which would at least provide a workaround 🤔

**Reply by JoviDeCroock** [reply]
I do think it might be a good time for that, alternatively, I personally have been seeing success doing `cache.resolve('Query', 'user', { id, username })` where we leverage the child-field of the Query rootType with the arguments to see if we resolve to a specific user

**Reply by arcanis** [reply]
Being able to get records of a certain type would be nice. I'm already doing it (by accessing the private state) to automatically invalidate lists when a new record is added/removed by a mutation.

**Reply by nandorojo** [reply]
@JoviDeCroock could you clarify the behavior of `cache.resolve('Query', 'user', { id, username }) `? Is this expected to work if you sometimes query by `id` and other times query by `username`?

**Reply by JoviDeCroock** [reply]
It is yes, because we create a cache-key for the type `Query` and the field `user` when we have queried this with username we can find it like that so a reasonable solution could be:

`if (id) cache.resolve('Query', 'user', { id }) else if username ...` 

The keys config itself only indicates what we will use to key a type so `User:id` in this case but does not impact how we have queried something and linked from the root to a type

**Reply by nandorojo** [reply]
In case anyone is wondering, this is what I did for a full code example:

```ts
resolvers: {
  Query: {
    organization(parent, args) {
      return { __typename: 'Organization', id: args.id }
    },
    organizationBySlug(parent, args, cache) {
      const key =
        cache.resolve('Query', 'organizationBySlug', {
          slug: args.slug,
        }) || null

      // @ts-expect-error broken types, expects string | { __typename, ...}
      return key
    },
  },
}
```
